File: Difficulty: Medium/M-Coloring Problem/mcoloring-problem.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int tc = scan.nextInt();

        while (tc-- > 0) {
            int N = scan.nextInt();
            int M = scan.nextInt();
            int E = scan.nextInt();

            boolean graph[][] = new boolean[N][N];

            for (int i = 0; i < E; i++) {
                int u = scan.nextInt() - 1;
                int v = scan.nextInt() - 1;
                graph[u][v] = true;
                graph[v][u] = true;
            }

            System.out.println(new solve().graphColoring(graph, M, N) ? 1 : 0);
        }
    }
}

// } Driver Code Ends


class solve {
    // Function to determine if graph can be coloured with at most M colours
    // such
    // that no two adjacent vertices of graph are coloured with same colour.
    boolean check(int src, int color, int[] colors, boolean[][] graph) {
        for (int i = 0; i < graph[src].length; i++) {
            if (graph[src][i] == true && colors[i] == color)
                return false;
        }
        return true;
    }
    
    boolean dfs(int src, boolean[][] graph, int[] colors, int m) {
        if (src >= graph.length) return true;
        for (int c = 1; c <= m; c++) {
            if (check(src, c, colors, graph)) {
                colors[src] = c;
                if (dfs(src+1, graph, colors, m))
                    return true;
                colors[src] = 0;
            }
        }
        return false;
    }
    
    public boolean graphColoring(boolean graph[][], int m, int n) {
        // Your code here
        int[] colors = new int[n];
        return dfs(0, graph, colors, m);
    }
}

File: Difficulty: Medium/M-Coloring Problem/mcoloring-problem.py
# User function Template for python3

from collections import defaultdict
class Solution:
    def graphColoring(self, V, edges, m):
        graph = defaultdict(list)
        colors = [-1] * V
        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        def isSafe(node, color):
            for neighbor in graph[node]:
                if colors[neighbor] == color:
                    return False
            return True
        
        def dfs(node):
            for color in range(m):
                if isSafe(node, color):
                    colors[node] = color
                    for child in graph[node]:
                        if colors[child] == -1:
                            if dfs(child) == False:
                                colors[node] = -1
                                return False
                    return True
            
            return False
        
        for i in range(V):
            if colors[i] == -1:
                if dfs(i) == False:
                    return False
        
        return True


#{ 
 # Driver Code Starts
def main():
    t = int(input())
    for _ in range(t):
        n = int(input())
        edges_input = list(map(int, input().split()))
        m = int(input())
        edges = []

        # Populate the edges list with edge pairs
        for i in range(0, len(edges_input), 2):
            edges.append((edges_input[i], edges_input[i + 1]))

        solution = Solution()
        print("true" if solution.graphColoring(n, edges, m) else
              "false")  # Call the graph coloring function


if __name__ == "__main__":
    main()

# } Driver Code Ends